src/main/java/net/sf/jabref/gui/exporter/SaveDatabaseAction.java

```diff
@@ -304,26 +315,58 @@ public void saveAs() throws Throwable {
             }
         }
 
-        File oldFile = panel.getBibDatabaseContext().getDatabaseFile().orElse(null);
-        panel.getBibDatabaseContext().setDatabaseFile(file);
+        BibDatabaseContext context = panel.getBibDatabaseContext();
+
+        if (context.getLocation() == DatabaseLocation.SHARED) {
+            // Generate an ID when saving a shared database
+            String databaseID = new BigInteger(128, new SecureRandom()).toString(32);
+            context.getDatabase().setDatabaseID(databaseID);
+            DBMSProcessor dbmsProcessor = context.getDBMSSynchronizer().getDBProcessor();
+            DBMSConnectionProperties properties = dbmsProcessor.getDBMSConnectionProperties();
+            // Save all properties dependent on the ID. This makes it possible to restore them.
+            new SharedDatabasePreferences(databaseID).putAllDBMSConnectionProperties(properties);
+        }
+
+        context.setDatabaseFile(file);
         Globals.prefs.put(JabRefPreferences.WORKING_DIRECTORY, file.getParent());
         runCommand();
         // If the operation failed, revert the file field and return:
         if (!success) {
-            panel.getBibDatabaseContext().setDatabaseFile(oldFile);
+            context.setDatabaseFile(context.getDatabaseFile().orElse(null));
```

Comment:
I don't understand this line (set/get from the same context), especially since you set the database file a few lines earlier.